components/embedded_form/embedded_form.jsx

```diff
@@ -0,0 +1,243 @@
+        actions: PropTypes.shape({
+            submitEmbeddedForm: PropTypes.func.isRequired,
+        }).isRequired,
+        emojiMap: PropTypes.object.isRequired,
+        postID: PropTypes.string.isRequired,
+        appID: PropTypes.string.isRequired,
+    };
+
+    constructor(props) {
+        super(props);
+
+        const values = {};
+        if (props.elements != null) {
+            props.elements.forEach((e) => {
+                if (e.type === 'bool') {
+                    values[e.name] =
+                        e.default === true ||
+                        String(e.default).toLowerCase() === 'true';
+                } else {
+                    values[e.name] = e.default || null;
+                }
+            });
+        }
+
+        this.state = {
+            values,
+            error: null,
+            errors: {},
+            submitting: false,
+        };
+    }
+
+    handleSubmit = async (e) => {
+        e.preventDefault();
+
+        const {elements} = this.props;
+        const values = this.state.values;
+        const errors = {};
+        if (elements) {
+            elements.forEach((elem) => {
+                const error = checkDialogElementForError(
+                    elem,
+                    values[elem.name],
+                );
+                if (error) {
+                    errors[elem.name] = (
+                        <FormattedMessage
+                            id={error.id}
+                            defaultMessage={error.defaultMessage}
+                            values={error.values}
+                        />
+                    );
+                }
+            });
+        }
+
+        this.setState({errors});
+
+        if (Object.keys(errors).length !== 0) {
+            return;
+        }
+
+        const {url, callbackId, state} = this.props;
+
+        const dialog = {
+            url,
+            callback_id: callbackId,
+            state,
+            submission: values,
+        };
+
+        this.setState({submitting: true});
+
+        const {data} = await this.props.actions.submitEmbeddedForm(
+            dialog,
+            this.props.postID,
+            this.props.appID,
+        );
+
+        this.setState({submitting: false});
+
+        //let hasErrors = false;
+
+        if (data) {
+            if (data.error) {
+                //hasErrors = true;
+                this.setState({error: data.error});
+            }
+
+            if (
+                data.errors &&
+                Object.keys(data.errors).length >= 0 &&
+                checkIfErrorsMatchElements(data.errors, elements)
+            ) {
+                //hasErrors = true;
+                this.setState({errors: data.errors});
+            }
+        }
+
+        // TODO handle non error data to update the form?
+    };
+
+    onChange = (name, value) => {
+        const values = {...this.state.values, [name]: value};
+        this.setState({values});
```

Comment:
`setState` calls can be batched and run asynchronously, so if you're setting state from itself, the preferred way to do that is
```suggestion
        this.setState((state) => {
            const values = {...state.values, [name]: value};
            return {values};
        });
```

Reply:
Regarding typescript, it seems the migration would be a bit more complicated, since there seem to be some discrepancies on redux. I may try at some other point.

Reply:
What sort of discrepencies? Are they problems with mattermost-redux that we need to track, or are they issues with redux itself? I know there's some issues with some of the mattermost-redux type definitions, so we could take note of anything that needs to be fixed.